File: funcscript/funcs/logic/EqualsFunction.cs
using funcscript.core;
using System;
using funcscript.model;

namespace funcscript.funcs.logic
{
    public class EqualsFunction : IFsFunction, IFsDref
    {
        public int MaxParsCount => 2;

        public CallType CallType => CallType.Infix;

        public string Symbol => "=";

        public int Precidence => 200;

        public object Evaluate(IFsDataProvider parent, IParameterList pars)
        {
            if (pars.Count != MaxParsCount)
                throw new error.TypeMismatchError($"{this.Symbol} function: Invalid parameter count. Expected {MaxParsCount}, but got {pars.Count}");

            var parBuilder = new CallRefBuilder(this, parent, pars);
            var par0 = parBuilder.GetParameter(0);
            var par1 = parBuilder.GetParameter(1);

            if (par0 is ValueReferenceDelegate || par1 is ValueReferenceDelegate)
                return parBuilder.CreateRef();

            return EvaluateInternal(par0, par1);
        }

        private object EvaluateInternal(object par0, object par1)
        {
            if (par0 == null && par1 == null)
                return true;

            if (par0 == null || par1 == null)
                return false;

            if (FuncScript.IsNumeric(par0) && FuncScript.IsNumeric(par1))
            {
                FuncScript.ConvertToCommonNumericType(par0, par1, out par0, out par1);
            }

            return par0?.GetType() == par1?.GetType() && par0.Equals(par1);
        }

        public object DrefEvaluate(IParameterList pars)
        {
            var par0 = FuncScript.Dref(pars.GetParameter(null, 0),false);
            var par1 = FuncScript.Dref(pars.GetParameter(null, 1),false);
            return EvaluateInternal(par0, par1);
        }

        public string ParName(int index)
        {
            switch(index)
            {
                case 0:
                    return "Left Value";
                case 1:
                    return "Right Value";
                default:
                    return "";
            }
        }
    }
}

File: js-port/fs-js/src/funcs/logic/equals-function.js
const { BaseFunction, CallType } = require('../../core/function-base');
const helpers = require('../helpers');

class EqualsFunction extends BaseFunction {
  constructor() {
    super();
    this.symbol = '=';
    this.callType = CallType.Infix;
  }

  get maxParameters() {
    return 2;
  }

  evaluate(provider, parameters) {
    const error = helpers.expectParamCount(this.symbol, parameters, this.maxParameters);
    if (error) {
      return error;
    }

    let left = helpers.ensureTyped(parameters.getParameter(provider, 0));
    let right = helpers.ensureTyped(parameters.getParameter(provider, 1));

    if (helpers.typeOf(left) === helpers.FSDataType.Null && helpers.typeOf(right) === helpers.FSDataType.Null) {
      return helpers.makeValue(helpers.FSDataType.Boolean, true);
    }
    if (helpers.typeOf(left) === helpers.FSDataType.Null || helpers.typeOf(right) === helpers.FSDataType.Null) {
      return helpers.makeValue(helpers.FSDataType.Boolean, false);
    }

    if (helpers.isNumeric(left) && helpers.isNumeric(right)) {
      [left, right] = helpers.convertToCommonNumericType(left, right);
    }

    if (helpers.typeOf(left) !== helpers.typeOf(right)) {
      return helpers.makeValue(helpers.FSDataType.Boolean, false);
    }

    return helpers.makeValue(helpers.FSDataType.Boolean, helpers.valueOf(left) === helpers.valueOf(right));
  }
}

module.exports = {
  EqualsFunction
};
